diagnostics-app/src/app/routine-v2/routine-v2-card/routine-v2-card.component.ts

```diff
@@ -0,0 +1,246 @@
+  @Input({required: true}) routineArgument!: RoutineV2Argument;
+
+  public RoutineV2State = RoutineV2State;
+
+  private _routineOutput?: RoutineV2FinishedInfoUnion;
+  // the output, which may be either routine result or user action.
+  private _output?: object | string;
+  // whether the routine has passed. Only shown if the routine is finished.
+  private _hasPassed?: boolean | undefined;
+  // the error message if any.
+  private _error?: string | undefined;
+  // the routine id associated with the card, possibly null if no routine is running.
+  private _uuid?: string | undefined;
+  // An integer between 0 ~ 100 detailing the percentage ran of the routine.
+  private _percentage = 0;
+  // the state of the underlying routine
+  private _routineState: RoutineV2State = RoutineV2State.FINISHED;
+
+  get error() {
+    return this._error;
+  }
+  get output() {
+    return this._output;
+  }
+  get percentage() {
+    return this._percentage;
+  }
+  get routineState() {
+    return this._routineState;
+  }
+  get hasPassed() {
+    return this._hasPassed;
+  }
+  get passedColor() {
+    if (this._hasPassed) {
+      return 'green';
+    }
+    return 'red';
+  }
+  get messageColor() {
+    if (
+      this._routineState === RoutineV2State.EXCEPTION ||
+      this._routineState === RoutineV2State.CANCELLED
+    ) {
+      return 'yellow';
+    }
+    return 'white';
+  }
+
+  constructor(
+    private routineV2Service: RoutineV2Service,
+    private changeDetectorRef: ChangeDetectorRef,
+  ) {}
+
+  ngOnInit(): void {
+    this.routineV2Service
+      .getSubject(this.routineArgument)
+      .then((res) => {
+        if (res.success && res.subject) {
+          res.subject.subscribe({
+            next: (event: RoutineV2Event) => {
+              this.handleEventResponse(event);
+            },
+          });
+        } else {
+          this._error = res.error;
+        }
+      })
+      .catch((err) => {
+        this._error = err.message;
+      });
+  }
+
+  ngOnDestroy(): void {
+    if (this._uuid) {
+      this.routineV2Service.CancelRoutine(this._uuid);
+    }
+  }
+
+  // this function is used for maintaining the attributes' original order
+  public originalOrder = (
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _a: KeyValue<number, string>,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _b: KeyValue<number, string>,
+  ): number => 0;
+
+  async handleEventResponse(event: RoutineV2Event) {
+    switch (event.eventCategory) {
+      case RoutineV2EventCategory.initialized: {
+        this._output = 'initialized';
+        this._routineOutput = undefined;
+        this._routineState = RoutineV2State.RUNNING;
+        const routineInitializedInfo = event.event as RoutineInitializedInfo;
+        if (this._uuid !== routineInitializedInfo.uuid) {
+          console.error('UUID not matched');
+          this._error = 'ERROR: uuid not matched';
+        }
```

Comment:
Should we wrap the duplicated codes as another function?

Reply:
Done.